sympy/core/mul.py

```diff
@@ -1248,15 +1248,25 @@ def _eval_is_zero(self):
         return zero
 
     def _eval_is_integer(self):
+        from sympy import fraction
+
         is_rational = self.is_rational
+        if is_rational is False:
+            return False
 
+        n, d = fraction(self)
+        if not is_rational:
```

Comment:
It is definitely better if we can avoid evaluating new objects as part of assumptions checking. This sort of thing can easily lead to infinite recursion.